src/VirtoCommerce.Platform.Caching/Redis/RedisPlatformMemoryCache.cs

```diff
@@ -1,47 +1,90 @@
 using System;
+using System.Collections.Generic;
+using Microsoft.ApplicationInsights;
 using Microsoft.Extensions.Caching.Memory;
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
 using Newtonsoft.Json;
-using Polly;
-using Polly.Retry;
 using StackExchange.Redis;
 using VirtoCommerce.Platform.Caching;
+using VirtoCommerce.Platform.Core.Caching;
 using VirtoCommerce.Platform.Core.Common;
 
 namespace VirtoCommerce.Platform.Redis
 {
     public class RedisPlatformMemoryCache : PlatformMemoryCache
     {
         private readonly ISubscriber _bus;
+        private readonly IMemoryCache _memoryCache;
+        private readonly CachingOptions _cachingOptions;
         private readonly RedisCachingOptions _redisCachingOptions;
         private readonly ILogger _log;
-        private readonly RetryPolicy _retryPolicy;
+        private readonly TelemetryClient _telemetryClient;
 
         private static readonly string _cacheId = Guid.NewGuid().ToString("N");
         
 
         public RedisPlatformMemoryCache(IMemoryCache memoryCache, IOptions<CachingOptions> options
+            , ConnectionMultiplexer connection
             , ISubscriber bus
+            , IOptions<CachingOptions> cachingOptions
             , IOptions<RedisCachingOptions> redisCachingOptions
             , ILogger<RedisPlatformMemoryCache> log
+            , TelemetryClient telemetryClient
             ) : base(memoryCache, options, log)
        {
+            _memoryCache = memoryCache;
             _log = log;
+            _telemetryClient = telemetryClient;
             _bus = bus;
 
+            _cachingOptions = cachingOptions.Value;
             _redisCachingOptions = redisCachingOptions.Value;
-            _bus.Unsubscribe(_redisCachingOptions.ChannelName);
-            _bus.Subscribe(_redisCachingOptions.ChannelName, OnMessage);
+
+            connection.ConnectionFailed += OnConnectionFailed;
+            connection.ConnectionRestored += OnConnectionRestored;
+
+            _bus.Subscribe(_redisCachingOptions.ChannelName, OnMessage, CommandFlags.FireAndForget);
 
             _log.LogInformation($"{nameof(RedisPlatformMemoryCache)}: subscribe to channel {_redisCachingOptions.ChannelName } current instance:{ _cacheId }");
+            _telemetryClient.TrackEvent("RedisSubscribed", new Dictionary<string, string>
+            {
+                {"channelName", _redisCachingOptions.ChannelName},
+                {"cacheId", _cacheId}
+            });
+        }
+
+        protected virtual void OnConnectionFailed(object? sender, ConnectionFailedEventArgs e)
```

Comment:
for what reason object type marked with a question mark at the end? I guess there is no difference so I suggest you remove it

Reply:
Already fixed. It was auto-generated this way because if non-nullable classes existence in C# 8